### ISSUE TITLE
[path_provider] Completely rethink the path_provider API

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#118712]
The API for `path_provider` has a number of significant issues:
- It was design before federation existed, so platform-specific paths (like `getExternalCacheDirectories`, which is highly Android-specific in practice) are mixed in with abstractions in the app-facing API.
- Some of the APIs were designed for Android APIs that have since been deprecated, and the underlying implementation changed, creating some confusing naming.
- The APIs are almost all pre-desktop, so tend to be mobile-centric.
- The desktop implementation of the documents directory (a widely-used API) is, essentially, [wrong](https://github.com/flutter/flutter/issues/77628), and we can't fix it without a breaking change.

The package's API would benefit from being redesigned from scratch. We should:
- Audit all of the existing APIs and open requests for new APIs.
- Make a matrix of all the APIs that it makes sense to support, and what platforms they make sense on.
- Decide what high-level abstractions we want core functionality for (in particular taking a look at the user-created vs. app-created-but-expensive vs app-created-and-easily purgeable, which we have as distinctions right now but in a somewhat platform-fragmented way, and not well named as abstractions).
- Create a two-tiered API system, with common abstractions in `path_provider` and platform-specific things in `path_provider_*`, with clear docs in the app-facing package about the existence of the latter, with links to implementation package readmes as examples.

Removing APIs from path_provider would be extremely disruptive as it's such a foundational plugin in the package ecosystem, but we can deprecated the old APIs and create a clean set of new APIs, like we did with `url_launcher`. That way we don't have to do it as a breaking change, so everyone in the ecosystem can switch to the new APIs gradually, without creating a lot of transitive version conflict problems.

**Comment by pattobrien** [comment]
When rethinking these APIs, would it also be possible to support standalone Dart applications? (e.g. `getApplicationSupportDirectory`)

**Comment by stuartmorgan** [comment]
@pattobrien Please see https://github.com/flutter/flutter/issues/122529. Also, the implementation and the client-facing API are largely orthogonal, so that's unrelated to this issue.

**Comment by dickermoshe** [comment]
@stuartmorgan 
You seem to be the one to talk to about `path_provider`.

I've ported path_provider over to use `ffi` for every platform.
My main motivation for this is that the APIs should be synchronous and it wont depend on Flutter.

I would like to know the best way of going about this.

I think we should start by removing each plugin at a time, starting with `windows` and `linux` which use ffi currently anyway.

path_provider is also missing tests for some platforms which is also needed.

We can make the methods synchronous on the next breaking release.

**Comment by stuartmorgan** [comment]
> My main motivation for this is that the APIs should be synchronous

https://github.com/flutter/flutter/issues/112009 would be the place to discuss that.

> and it wont depend on Flutter.

Supporting non-Flutter use cases is [not a goal for the plugin](https://github.com/flutter/flutter/issues/144400) (and that wouldn't be related to this issue regardless).

**Comment by Gustl22** [comment]
Agreed, the returned paths differ widely across the platforms. E.g. there is no unified way of accessing an app-specific directory which can be accessed by the user: https://github.com/flutter/flutter/issues/118571#issuecomment-3506876530
